Extend the user guide, and link to it from the README

david-a-wheeler opened the pull request:
This extends the user guide to walk through the simple example ( 2 + 2 ) = 4, as well as significantly expanding the functionality it covers.

expln commented on docs/guide.md:
```diff
@@ -11,15 +11,194 @@ More information on metamath-lamp is available at the
+  until the goal is completely proven.
+  You can do it bottom-up, top-down, middle-out, whatever makes sense to you.
+* Copy the compressed proof of the goal into the clipboard.
+  You can do this my selecting the green checkmark next to the goal
```
Maybe it will be better "by clicking"?

david-a-wheeler replied:
Agreed!

expln commented on docs/guide.md:
```diff
@@ -11,15 +11,194 @@ More information on metamath-lamp is available at the
+would simply reuse the existing proof.
+
+> Select Source type "Web", Alias "set.mm:latest"; after confirmation this
+> load load the given database.
```
typo "load load" -> "loads"

expln commented on docs/guide.md:
```diff
@@ -11,15 +11,194 @@ More information on metamath-lamp is available at the
+> `|- 2 = ( 1 + 1 )`
+
+Because 2 is just `( 1 + 1 )` we can now show that
+is `( 2 + ( 1 + 1 ) )` is equal to `( 2 + 2 )`.
```
Looks like a typo, the first "is" on this line is redundant. But I may be missing something.

expln commented on docs/guide.md:
```diff
@@ -11,15 +11,194 @@ More information on metamath-lamp is available at the
+
+Because 2 is just `( 1 + 1 )` we can now show that
+is `( 2 + ( 1 + 1 ) )` is equal to `( 2 + 2 )`.
+We'll add the claim and let metamath-lamp proof that.
```
Looks like a typo "proof that" -> "prove that". ChatGPT agrees with me, but again I may be missing something, I am not a native English speaker.

expln commented on docs/guide.md:
```diff
@@ -117,8 +298,154 @@ icons and meanings:
+
+The basic information about the proof are the proof's description,
+variables, and disjoints. Click on the text area to the right of those
+fields to edit this infomration. Here's some about those fields:
```
"Click on the text area to the right of those fields to edit this infomration."

Typo "infomration" -> "information"

This works for variables and disjoints, but not for description. Description may contain links and some useful text. Users will be clicking links and copying parts of the description. So they could end up with the description in "edit" mode rather then with a part of description selected if a single click would enable "edit" mode similar to variables and disjoints. So I decided to change how to begin editing for the description - you have to do Alt+<left click>. Also understanding that this adds a bit of complexity to the UI, I enabled another feature - you may click the word "Description" itself, this will also enter the "edit" mode. The same is true for variables and disjoints. So now it is better to click the word "Variables" rather then positioning a mouse pointer in the small area next to it.

I would change this statement to something like that:
"Click on the section name one of "Description", "Variables" or "Disjoints" to edit corresponding section."

david-a-wheeler replied:
I'll change the text to what it does, but I think this inconsistency is *not* a good idea. Left-click should edit; alt+left click (or long click) should select. Exceptions create UI problems.

expln commented on docs/guide.md:
```diff
@@ -117,8 +298,154 @@ icons and meanings:
+  Conventionally this includes, at its end, a statement like
+  "(Contributed by NAME, DD-MMM-YYYY)" where DD-MMM-YYYY is the date
+  the proof was completed and MMM is the 3-letter English name
+  of the month.
```
Actually the description is not included into the final proof. And mm-lamp doesn't support these formatting rules. So I would not mention these formatting rules in this guide. Moreover this is not needed for newcomers :) If somebody wants to contribute to set.mm then probably it is better to include a reference to a corresponding document describing rules how to contribute to set.mm.

I implemented different formatting rules for the description. You may see them by entering "edit" mode for the description and clicking the question mark button to the right. My intention is that people may share some small proofs via url and use the description to include information like "_Check the proof below. See how cool it looks when we introduce such and such definitions, constants etc._"

So, I would change this text to:

`* Description: May contain some useful remarks and comments you want to give to anyone reading your proof when you share it with other people. The description supports a basic formatting rules. You may see them by entering "edit" mode for the description and clicking the question mark button to the right.`

david-a-wheeler replied:
Ok, I'll move most of this text to an HTML comment where it won't be seen.

I think this is a mistake - the description should be generated as a comment just before the proof. It's okay to support HTML, so does metamath-exe, just include <html> at the beginning if anything requires it. It'd be easier to round-trip if description matched the database.

expln replied:
There are few reasons why I decided not to support Metamath formatting and why I chose the current implementation:

1) I think that if anyone uses this tool for creating real proofs to add them to one of the official databases, most probably it will be more convenient for them to use their favorite text editor. That should not be difficult. mm-lamp prepares everything ready to copy-paste to a database and only the comment remains to be added. It is more convenient to prepare it in a normal text editor which supports different hot keys, auto indentation, highlighting and so on.

2) I don't want to spend time on developing a new parser and formatter (at least now). Current implementation is not html, but it is very similar to html and easy to transform it to html if needed. I spent less than a day to implement it from scratch and still it remains very extensible. It uses an xml parser embedded into browsers, that's why it was easy to implement it and I hope it (the xml parser) is bug free.

3) I am not ready to use any existing rich markup library due to possible security issues. Embedding raw html is even more dangerous. Imagine a scenario when someone sends a url saying "I formalized a very interesting proof, check it in mm-lamp by this url", but when you open the url it does some malicious actions :) Current implementation is very strict in terms of what is allowed to use for formatting, so I think it doesn't have any of those security issues.

I am not against mentioning set.mm specific formatting rules in this guide, but I would like to mention my initial intention for the description field in the first place, which is sharing ideas, asking questions, and maybe creating some small interactive tutorials. Your idea to place editor state into URL is great, I never imagined anything like that before you proposed it. But then I immediately understood that this approach maybe used to help sharing knowledge :)

david-a-wheeler replied:
In the longer term I think "Description" should go into the database, but it doesn't do that today. So I'll file an issue where that can be discussed, and make the "Description" discussion short for now.

expln commented on docs/guide.md:
```diff
@@ -117,8 +298,154 @@ icons and meanings:
+  and show that they are of the correct types.
+* Id: This is an arbitrary id for this statement.
+  By convention the goal is labelled "qed" or the name of the statement
+  to be proved, and the hypotheses are labelled "hNUMBER" where NUMBER is a
```
Naming "h1", "h2", ... actually will not always work. For example, I've just checked: when loading set.mm and naming a hypothesis "h1" I am getting an error `Cannot reuse label 'h1' [1].` 

I implemented stricter rules for hypothesis naming, because previously I was getting errors in metamath-exe similar to that one:
```
?Error on line 732616 of file "set-check-new-proofs.mm" at statement 203351,
    label "1", type "$e":
            1 $e |- ch $.
            ^
    This label has the same name as the math token declared on line 108187 of file
```

To be honest, I would not mention naming conventions in this guide because this is not enforced by the tool, and for newcomers this looks more like a restriction rather then something intended to simplify work. Also labels of hypothesis will eventually become labels of essential statements when generating a compressed proof. So it makes sense to name them how you want them to appear in set.mm (if you are planning to eventually include the proof to set.mm).

david-a-wheeler replied:
Oh, *that's* how it works in metamath-lamp! Thanks so much for seeting me straight. That is quite reasonable, it's just different from mmj2.

The naming conventions will affect whether or not the proof can be added, so I think we have to say *something*. But I'll try to make what's said correct this time :-).

expln commented on docs/guide.md:
```diff
@@ -11,15 +11,194 @@ More information on metamath-lamp is available at the
+> and Enter (Return) to save the result.
+> Be sure to surround each symbol by at least one space, and the
+> parentheses are not optional.
+> Finally, select the statement number (1) and change its name to "qed".
```
I would explicitly specify that mm-lamp doesn't enforce this naming, so newcomers don't have questions "why qed?", "what does qed mean?". I would add:

```
Renaming the target statement (as well as any other statement) is not required and you may not do this and just continue with whatever label metamath-lamp suggested. But renaming some of the statements (especially the target one) makes them easier to distinguish and eventually this name will appear in the final proof.
```

david-a-wheeler replied:
Sure!

expln commented on docs/guide.md:
```diff
@@ -11,15 +11,194 @@ More information on metamath-lamp is available at the
+> and press Return (Enter).
+> Now press unify (the multiple-connected dots symbol).
+> This will show a green checkmark next to our new statement.
+
```
This is the point where moving "qed" to the bottom comes into place :) I would modify the beginning of this instruction to eliminate moving qed to the end in further steps:

```
Select the checkbox to the left of the "qed" statement. Then select "+" (add new statement). As a result a new statement gets inserted before "qed". Enter the new statement
> `|- ( 3 + 1 ) = ( ( 2 + 1 ) + 1 )`
> and press Return (Enter). ...
```

david-a-wheeler replied:
Good idea.

expln commented on docs/guide.md:
```diff
@@ -11,15 +11,194 @@ More information on metamath-lamp is available at the
+> `|- ( 3 + 1 ) = ( 2 + ( 1 + 1 ) )`
+> along with two new statements, `1 e. CC` (1 is a complex number)
+> and `2 e. CC` (2 is a complex number).
+
```
Here again I would instruct to select and unselect qed:

```
> THIS IS ADDED: Select the checkbox to the left of the "qed" statement.
> Select "+" (add new statement). Enter the new statement
> `|- ( ( 2 + 1 ) + 1 ) = ( 2 + ( 1 + 1 ) )`
> and press Return (Enter).
> THIS IS ADDED: Unselect the checkbox to the left of the "qed" statement.
> We could try to simply Unify, but this is more than a trivial unification -
> ...
```

david-a-wheeler replied:
Yes, will do. However, this constant selecting before adding suggests something is wrong with "+", the default should "do the right thing in most cases" and that doesn't seem to be the case. That's a longer-term issue though, the guide should document current behavior.

expln commented on docs/guide.md:
```diff
@@ -11,15 +11,194 @@ More information on metamath-lamp is available at the
+> you'll see that one of the first options uses `oveeq2i` and completely
+> proves it, so select that and "apply selected".
+
+At this point metamath-lamp has noticed that it can prove every statement,
```
I followed all the steps including my suggestions to add "selecting/unselecting qed" and bottom-up proving was not required at the end. A simple unification was able to complete the proof. Though I am not sure if this is because of my addition of "selecting/unselecting qed". I need to check it once again.

david-a-wheeler replied:
Huh, you're right! I've modified the text.

expln reviewed (commented):
This is a lot of work! Thanks David! I reviewed and left few comments. Many of them are simple typos. However, some of them are touching more complex things related to my vision of how this guide may be used and understood by newcomers.

david-a-wheeler commented:
I'll post soon an update with the fixes.

david-a-wheeler commented:
Here's an update, I think I addressed all your recommendations.

I also changed how I proved 2+2=4 a little bit. I still expand 4=3+1, then 3+1=(2+1)+1. However, the change is that I immediately "clean up" by showing that "4=(2+1)+1". This is probably a better general practice when proving forwards, and it makes the explanation clearer.

expln reviewed (commented):
Great! Thank you!

The new sequence of proof steps especially with detailed explanations of why we are doing something is much more clear for me now. Introduction of duplication action demonstrates one more feature and simultaneously removes one "select/unselect qed". Moving bottom-up proof to the end makes an impression of gradual introduction of new features which in general should make reading of this example more "memorable" (as for me).

I am not sure how you are going to proceed with updating of this guide. But now it looks complete in terms of introduction and the first example. I would appreciate it if you could fix new comments (mostly typos) and I then would complete the PR, and then you could create a new PR for further changes. This will allow me to review new changes more conveniently.

Thanks a lot!

expln commented on docs/guide.md:
```diff
@@ -115,10 +363,232 @@ icons and meanings:
+  All other statements are typically
+  consecutive integers starting with 1, though they don't need to be;
+  you can use any sequence of alphanumerics for an id as long as it's
+  not already in the database.
```
Please disregard this comment. I think I mistook. I will check it later and update you if needed. Obviously it should not be possible to have a hypothesis in the database and a statement in the editor with same label.

david-a-wheeler commented on docs/guide.md:
```diff
@@ -11,15 +11,261 @@ More information on metamath-lamp is available at the
+and eventually this name will appear in the final proof.
+
+> Select the statement number (1) using the checkbox to the left of it.
+> Change its name to "qed" and press Enter (Return).
```
Agreed. It's changed in my local version, which I will soon send.

david-a-wheeler commented on docs/guide.md:
```diff
@@ -115,10 +363,232 @@ icons and meanings:
+  You should give your proof's goal the id of what you intend to name it.
+  Consider naming your goal "qed" if you don't know what name to use.
+  Each hypothesis needs to have a unique id that isn't already in the
+  database are using.
```
I'll reword that slightly.

david-a-wheeler commented on docs/guide.md:
```diff
@@ -11,15 +11,261 @@ More information on metamath-lamp is available at the
+> and press Enter (Return).
+> Unselect the "qed" statement.
+> As an experiment, select Unify with no statement selected;
+> you'll see that in this case nothing happens.
```
I like this idea! However, I want to make it sure it works in context (as expected) by re-running the whole thing from scratch. I'm also afraid of making too many more changes in a still-unmerged PR, as that gets harder to follow.

So I've added this proposed change as issue #37. Once this guide is merged (as PR #28 ), we can make that change separately.

david-a-wheeler commented:
> I am not sure how you are going to proceed with updating of this guide.

I have some thoughts for next steps after this PR is merged. Here is my current thinking:

Make it clear that this document is a "guide" that contains both a "user's guide/tutorial" and a "reference guide" (which walks through the UI and explains things in detail).

In the document itself, the first major section will be the user's guide which walks through a few examples, using `2+2=4` as the template. I think `2+2=4` should be the first one; it has no variables or disjoints, and involves knowledge that a grade schooler would understand. I intentionally didn't include partial selections in 2p2e4 so it doesn't introduce *too* many new concepts.

I'd like to add reccot to the set of examples, maybe the second or third example. Hopefully we can "steal" the text you've already written.

Currently there's a section for "more examples" at the end of the document, but I now think that all the examples should be at the beginning. That way, a new user can just sit down with the document, go through the worked examples (preferably by doing them), and understand how to use the tool. Once the tool has had some changes (as you're planning), I'm thinking about making a video from the 2p2e4 example.

The rest of the document is already the reference guide. There are a few sections not filled in ("TODO") that need filling in. In some cases they're TODO because I don't know the answers. In particular, I'm not sure what some of the proving bottom-up options are exactly about. Feel free to fill in any TODOs, and I can comment or suggest further improvements.

expln commented:
Ok. Thanks David. I merged the PR. I will read and reply to your comment a bit later.

expln commented:
> I think 2+2=4 should be the first one

I agree, it seems to be an excellent first example.

> Hopefully we can "steal" the text you've already written.

Not sure what text you mean in the context of reccot example. All my two videos about reccot were with no sound. But if you mean the sequence of steps shown in the video, then yes, please do :)

> Feel free to fill in any TODOs, and I can comment or suggest further improvements.

Sure, will do!

icecream17 commented on docs/guide.md:
```diff
@@ -11,15 +11,261 @@ More information on metamath-lamp is available at the
+There are many different ways to create a proof, including top-down,
+bottom-up, or even middle-out. Metamath-lamp supports them all.
+Also, note that there are often many proofs for the same true statement.
+Here we're going to show one way to do it, as an example.
```
In the context of the later changes, for clarity, I think this is "top-down". Or is "top-down" a different concept from the statement you want to prove being the last statement?

david-a-wheeler replied:
Maybe we should say "forwards toward goal" and "backwards from goal" for clarity.

david-a-wheeler replied:
This pr is closed source I missed the comment. Ill create a new issue to track it.

david-a-wheeler replied:
Okay, I'm about to create a pull request that changes the terminology, hopefuly that will eliminate the confusion.
